feat(loader): do not depend on flow-cli for feedback

SaveTheRbtz opened the pull request:
Upgraded blockRef code to a generic API-based follower so that we can wait until transaction is finished inline in the worker goroutine.  This removes the dependency on the `/dev/shm` and hence I've enabled transaction tracking by default.

If you like the approach, I can extend it to the rest of transaction types in this or following diff.  Left the `feedback-enabled` option in place to prevent cli breakage (but flipped the default.)

Followup fix for: dapperlabs/flow-go#6089

SaveTheRbtz commented on integration/utils/contLoadGenerator.go:
```diff
@@ -462,130 +436,68 @@ func (lg *ContLoadGenerator) sendAddKeyTx(workerID int) {
-			if (elapsed >= elapsedSuspicious) && (elapsed < (1 + elapsedSuspicious)) {
-				// come here to capture transactions which mysteriously have had no feedback after 30 seconds
-				// note: this should only be output once as its age reaches and passes through a certain amount of seconds...
-				lg.log.Warn().Msgf("workerID=%d address %x account %d tx_id=%s has been happening for %f seconds and counting", workerID, acc.address.Bytes(), acc.i, tx_id, elapsed)
```
I've added a debug print on `slowTransactionThreshold`, in followup diffs I'll add a metric for that.

SaveTheRbtz commented on integration/utils/contLoadGenerator.go:
```diff
@@ -462,130 +436,68 @@ func (lg *ContLoadGenerator) sendAddKeyTx(workerID int) {
-			// come here to have workerID 0 keep track of lowest number of available accounts in channel
-			if len(lg.availableAccounts) < lg.availableAccountsLo {
-				lg.availableAccountsLo = len(lg.availableAccounts)
-				lg.log.Debug().Msgf("workerID=%d discovered new availableAccountsLo=%d of numberOfAccounts=%d", workerID, lg.availableAccountsLo, lg.numberOfAccounts)
```
Added `txsInProgress` counter to the log, will also eventually add a corresponding metric to prometheus.

simonhf replied:
After looking at the business logic for the new implementation then I'm worried that the new implementation / counter might be misleading under certain circumstances.

Firstly: The new code replaces the lowest number of available accounts with the highest number of TXs in flight. So the opposite. Which is okay in principle, but I'm not sure what the motivation is switching. With the number of available accounts then this is perhaps easier for the reader / user to process because it should always be > 0. However, with the highest number of TXs in flight then this is perhaps more confusing to the reader / user. Why? Because the reader / user has to do more math to figure out what the ceiling is (TPS * account multiplier) that this stat gets close to, or? And when comparing multiple load tests, it's therefore easier to compare the lowest number of available accounts against zero, than doing the extra math to figure out how close the highest number of TXs in flight get to each ceiling, or?

Secondly: I'm not sure the highest number of TXs in flight is calculated accurately. Why? Consider this example:
* Ceiling is 5,000 TX and TPS is 100.
* Last block received contains 0 TX and shows 4,950 TX in flight at second n+0.
* Second n+1 tries to send 100 TPS but can only send 50; adds 50 to 4,950 = 5,000.
* Next block received contains 200 TX and shows 4,800 TX in flight at second n+2; subtracts 200 from 5,000 = 4,800.

In this example, the loader accounts would be exhausted but the TX in flight count would mislead the reader / user into thinking that the ceiling was never reached... when it in-fact was, or?

The problem is that the sending of TXs and block following are down asynchronously and both in batches -- batches of sending TX, and blocks with batches of TXs finalized -- but the output log is only done after receiving the latest block. Therefore, in the meantime, the atomic counters can get closer to the ceiling than actually reported, or?

My suggestion #1 is to introduce e.g. `txsInProgressHi` which always reports the highest number of TXs in flight so as to avoid the async math reporting issue above.

My suggestion #2 is to switch back to the opposite counter e.g. `availableAccountsLo` so it's easier to glance at this stat and see how close it gets to zero regardless of the TPS or accounts multiplier settings.

My suggestion #3 is to perhaps show both counters, `availableAccounts` (goes up and down using the atomic math and is effectively a snap shot value at the time the next block is finalized) and `availableAccountsLo` (only goes down and always shows the lowest value reached at any point so far)?

SaveTheRbtz replied:
The main reason for counting inflight transactions instead of the available accounts is because this layer of abstraction does not (and likely should not) know about the total number of accounts.  We should not care about a specific number here since if we the code upstream runs out of accounts to use it screams VERY loudly.

In the future diffs I'll add a counter that in a separate goroutine will emit a gauge of len(accounts) and a separate count on failure to get a new account from the channel.

simonhf replied:
> if we the code upstream runs out of accounts to use it screams VERY loudly

It's true, but even if it never screams and never runs out, then I'd like to know -- preferably by grepping the logs -- how much head room the load test had, i.e. how "healthy" it was running / how close it came to account exhaustion :-)

SaveTheRbtz commented on integration/utils/follower.go:
```diff
@@ -0,0 +1,158 @@
+			Uint64("height", block.Height).
+			Int("numCollections", len(block.CollectionGuarantees[:])).
+			Int("numSeals", len(block.Seals)).
+			Msg("new block found")
```
added timings and number of transactions to the log.

SaveTheRbtz commented on integration/loader/main.go:
```diff
@@ -38,7 +38,7 @@ func main() {
 	metricport := flag.Uint("metricport", 8080, "port for /metrics endpoint")
 	profilerEnabled := flag.Bool("profiler-enabled", false, "whether to enable the auto-profiler")
-	feedbackEnabled := flag.Bool("feedback-enabled", false, "whether to enable feedback / transaction tracking before account reuse (to avoid sequence number mismatch errors during transaction execution)")
+	_ = flag.Bool("feedback-enabled", true, "(deprecated) enabled by default")
```
left the flag but switched the default.

SaveTheRbtz commented on integration/utils/contLoadGenerator.go:
```diff
@@ -74,15 +68,8 @@ func NewContLoadGenerator(
-		accountCreationBatchSize = 250 // due to the bigger multiplier we enlarge accountCreationBatchSize to create the accounts faster
-	}
-	numberOfAccounts := tps * multiplier // 1 second per block, factor 10 for delays to prevent sequence number collisions
+	numberOfAccounts := tps * 20 // 1 second per block, factor 20 for delays to prevent sequence number collisions
```
Added a flag for a configurable multiplier w/ a default of 50.Added a flag for a configurable multiplier w/ a default of 50.

simonhf commented on integration/go.mod:
```diff
@@ -21,6 +21,7 @@ require (
 	github.com/plus3it/gorecurcopy v0.0.1
 	github.com/rs/zerolog v1.26.1
 	github.com/stretchr/testify v1.7.1
+	go.uber.org/atomic v1.9.0
```
Just out of interest, why use the Uber atomic package [1] instead of the built in atomic package?

I noticed it's used a lot more in flow-go, but not sure why... :-)

```
$ find . -type f | egrep "\.go" | xargs egrep "sync/atomic" | wc -l
7

$ find . -type f | egrep "\.go" | xargs egrep "go.uber.org/atomic" | wc -l
25
```

[1] https://pkg.go.dev/go.uber.org/atomic
[2] https://pkg.go.dev/sync/atomic

SaveTheRbtz replied:
This is mostly syntax sugar so one does not forget to use atomic methods on a sync/atomic variables.

simonhf commented on integration/utils/follower.go:
```diff
@@ -0,0 +1,231 @@
+			}
+			for _, tx := range col.TransactionIDs {
+				totalTxs++
+				if ch, loaded := f.txToChan.LoadAndDelete(tx.Hex()); loaded {
```
Suggestion: Having a counter for TX IDs discovered in blocks which are unknown? Why? Let's say multiple loaders are used concurrently, or a loader and additional ways of sending TXs... it would be nice to account for 'unknown' TXs in the log line below, or?

SaveTheRbtz commented on integration/utils/follower.go:
```diff
@@ -0,0 +1,231 @@
+		f.logger.Debug().
+			Hex("blockID", block.ID.Bytes()).
+			Dur("timeSinceLastBlock", time.Since(lastBlockTime)).
+			Dur("duration", time.Since(blockResolutionStart)).
```
The logic behind this was that getting transactions from the block involves `GetBlockByHeight` and N*`GetCollection` calls, each is O(10ms) which is way longer than SyncMap can take even on a pretty congested mutex.  That said, just in case we are concerned about the `GetBlockByHeight` call specifically, i've added a separate counter for this.

SaveTheRbtz commented:
Added log for available transactions.  Now we log number of available transactions during:
* getting a free account, at trace level
* "too long" warning, at warning level

Plus the inProgress warning periodically at info level.

SaveTheRbtz commented:
resolved merge conflicts and removed an accidental change to `numberOfKeysToAdd`

simonhf commented:
> Added log for available transactions ... getting a free account, at trace level

Is that this trace code log statement? Running in trace mode is very noisy and usually I wouldn't do that unless I was debugging. It would be nice to run in a less noisy mode but never the less see when the lowest available accounts occur without having a tsunami of logs :-)

```
	log.Trace().
		Int("availableAccounts", len(lg.availableAccounts)).
		Msgf("getting next available account")
```

Suggestion: Why not just implement the `availableAccountsLo` struct member like in the existing code? Why? It results in relatively few lines being output and therefore no need to run in trace mode to get a feeling for the account availability head room at runtime :-)

SaveTheRbtz commented:
My main concern here is that we are re-implementing various aggregation and counter functions of prometheus in the code.  How about just a prometheus Gauge (or a histogram) there?

simonhf commented:
> My main concern here is that we are re-implementing various aggregation and counter functions of prometheus in the code. How about just a prometheus Gauge (or a histogram) there?

Thanks for the comment and the perspective. It's true, and my guess is that the re-implementation is 3 or 4 lines of code. But isn't it worth it to avoid all sort of wasted manual time at run-time or after run-time, e.g. for scenarios like as follows:

Let's say I run 10 localnet load tests back to back over night, each with a differing TPS. The next morning I can do a single grep across the logs from all 10 load tests to discover the accounts available head room for each run. Or, create a report automatically with a single line of code showing the same stat. Or, during a load test, automatically shut down the load test if it doesn't meet certain criteria and is therefore not a clean run because it ran out of accounts. Or, if I want to glance at a run's progress while it's running -- e.g. to check its health and status -- I can look at the TMUX output and see all relevant stats in real-time.

How to achieve the same results with prometheus / grafana?

Find the head room the next morning: Presumably the prometheus docker instance which is started and stopped 10 times would have to be set up to remember its data and so it would get bigger and bigger over time, thus causing extra 'gardening' work to manually manage it when it gets too big? Plus the next morning I'd have to manually restart the prometheus / grafana docker instance in order to make the accounts available data available for access? Then manually fiddle with the grafana UI time range to show the 10 runs from the night before? Also, if I want to record these results in a notion page then I have to copy and paste a graph which is more manual work than copy and pasting some lines of grep output? Or is there a way to access prometheus from the command line -- e.g. before the run is finished? -- and ask prometheus programmatically to show the lowest account available head room?

Automatically abort the localnet load test if not a clean run, e.g. if accounts availability goes below a certain threshold? Presumably prometheus / grafana has some kind of notification mechanism? Does it make sense to use this to feed back into the localnet load test automation scripts to trigger a run shutdown with a log message to explain why? And how would the implementation work?

Glance at a run's progress while it's running: By only feeding the stats to prometheus instead of TMUX output, then the automation scripts would need to somehow auto open (and close?) a prometheus / grafana browser window. This might be possible on OSX where the loader results are less accurate, but everything works locally, but how to make this work on remote boxes such as GCP? Even if it were to work, then each prometheus / grafana stat to glance at would have to be in its separate pane. This might work, but we could easily run out of panes compared to the density of stats to glance at via TMUX output, or? Sure, one could scroll down the panes, but this is then more manual work scrolling around... Also, the amount of glancing would double because it'd be necessary to glance at the TMUX output and the grafana browser page?

I'm not trying to be difficult and just suggesting that prometheus / grafana might be a good tool for after the event "manual" (clicking and scrolling here and there) diagnosis, but not so good for automation? Therefore, I'm suggesting to keep prometheus / grafana as a useful tool but not try to use it as a hammer for all types of tasks? Disclaimer: I'm not a prometheus / grafana expert and there may well be easy and elegant solutions to the above use cases, which I'd also like to learn about :-)

Alternative suggestion to avoid logging and avoid prometheus / grafana: Consider using binstat [1] to report things like accounts available head room? binstat is potentially lower overhead than prometheus / grafana, does not mess with the log output, but never the less can be grepped, and its output can be used to trigger e.g. a run shutdown due to some kind of out of bounds condition?

Another alternative suggestion: Or we could just say that we don't care about wasting time manually clicking and scrolling or typing in grafana URLs etc... :-)

[1] https://github.com/onflow/flow-go/tree/master/utils/binstat

SaveTheRbtz commented:
I do disagree that this 4 lines of code are needed since we not only need to write them, but also test and support them.  Meanwhile a prometheus metric would be accessible on via `/metrics` of the stats port of the `loader` binary and it is literally 1 http request away.  That said, I'm OK with a compromise if it makes your life easier on the testing side.  

Rebased on top of master. Added a commit that reverts removal of `availableAccountsLo`.

simonhf commented:
@SaveTheRbtz @gomisha FYI I tried to run the loader in this PR with a modified localnet loader automation, but the automation failed. Why? Turns out that this previous PR [1] changed localnet to use loki for docker logs, whereas the automation scripts tail the docker logs on Linux in real-time to scan for e.g. errors and transaction execution failures, and to take a copy of the logs and compress them after each load test.

I'm wondering if it's possible to keep loki and never the less have this functionality? Any ideas? If no ideas then I will research this further on Monday...

[1] https://github.com/onflow/flow-go/pull/2510#issuecomment-1152839284

SaveTheRbtz commented:
Given that the automation is already broken by me in #2510 should we just proceed with this diff to unblock the rest of the load changes?

Kay-Zee commented:
Definitely not great that we broke @simonhf 's automation.

If you agree now with the premise of this PR @simonhf, then we should just merge it, and then have @SaveTheRbtz help you fix the automation after the fact (i.e. after all the blocked PRs are in). Thoughts?

simonhf commented:
@SaveTheRbtz offered some useful suggestions here [1] over the weekend which I intend to try out.

Since I'm blocked due to the automation being broken then I will work on un-breaking it today :-) Once un-broken, hopefully I'll be able to run it together with the loader version in this PR and give proper feedback and give a thumbs up? :-)

@key-Zee or is there a rush to merge the latest loader *today* even though I haven't been able to test it yet?

[1] https://github.com/onflow/flow-go/pull/2510#issuecomment-1152859152

SaveTheRbtz commented:
@simonhf have you managed to verify whether feedback mechanism works? (I would assume that even without spans the TPS data should be available.). If yes, should we land this diff and fix trace downloads in followup ones?

simonhf commented:
> have you managed to verify whether feedback mechanism works?

I'll take another look now that promtail is providing the loki logging workaround / fix :-)

simonhf commented:
* When testing with the new loader at 220 TPS, it appears to be much slower at pre-creating accounts than the old loader.
* It pre-creates more total accounts by default which is okay, but the time to create a batch of 250 accounts is longer too :-(

```
$ # old loader
$ cat run-600-loader.sh.log | egrep "creating and funding 250 accounts" | perl -lane 'm~^([\d\.]+)~; ($ts)=($1); if(defined $ts_old){ printf qq[- created 250 accounts in %9.6f seconds\n], $ts - $ts_old; } $ts_old = $ts;'
- created 250 accounts in 10.178849 seconds
- created 250 accounts in  9.091365 seconds
- created 250 accounts in 10.086919 seconds
- created 250 accounts in 10.092465 seconds
- created 250 accounts in 10.075681 seconds
- created 250 accounts in  9.097430 seconds
- created 250 accounts in  9.076952 seconds
- created 250 accounts in 10.071994 seconds
- created 250 accounts in  9.067893 seconds
- created 250 accounts in  9.071017 seconds
- created 250 accounts in 10.072174 seconds
- created 250 accounts in  9.069186 seconds
- created 250 accounts in 10.067833 seconds
- created 250 accounts in 10.077978 seconds
- created 250 accounts in 10.072676 seconds
- created 250 accounts in  9.079092 seconds
- created 250 accounts in 10.077267 seconds
- created 250 accounts in 10.073799 seconds
- created 250 accounts in  9.076619 seconds
- created 250 accounts in 10.073898 seconds
- created 250 accounts in  9.068252 seconds
- created 250 accounts in  9.098619 seconds
- created 250 accounts in  9.072330 seconds
- created 250 accounts in  9.079367 seconds
- created 250 accounts in  9.076400 seconds
```

```
$ # new loader
$ cat run-600-loader.sh.log | egrep "creating accounts" | perl -lane 'm~^([\d\.]+)~; ($ts)=($1); if(defined $ts_old){ printf qq[- created 250 accounts in %9.6f seconds\n], $ts - $ts_old; } $ts_old = $ts;'
- created 250 accounts in 16.539672 seconds
- created 250 accounts in 16.433452 seconds
- created 250 accounts in 17.415589 seconds
- created 250 accounts in 17.457562 seconds
- created 250 accounts in 17.354745 seconds
- created 250 accounts in 17.466239 seconds
- created 250 accounts in 17.424409 seconds
- created 250 accounts in 17.486329 seconds
- created 250 accounts in 17.412508 seconds
- created 250 accounts in 16.422745 seconds
- created 250 accounts in 17.419908 seconds
- created 250 accounts in 16.408091 seconds
- created 250 accounts in 17.439702 seconds
- created 250 accounts in 16.393792 seconds
- created 250 accounts in 17.440429 seconds
- created 250 accounts in 17.431484 seconds
- created 250 accounts in 17.421166 seconds
- created 250 accounts in 18.458669 seconds
- created 250 accounts in 16.414609 seconds
- created 250 accounts in 17.398953 seconds
- created 250 accounts in 16.387531 seconds
- created 250 accounts in 17.408620 seconds
- created 250 accounts in 16.403804 seconds
- created 250 accounts in 17.411264 seconds
- created 250 accounts in 15.372801 seconds
- created 250 accounts in 17.397953 seconds
- created 250 accounts in 16.397188 seconds
- created 250 accounts in 17.406631 seconds
- created 250 accounts in 16.424687 seconds
- created 250 accounts in 16.355350 seconds
- created 250 accounts in 17.486252 seconds
- created 250 accounts in 17.473318 seconds
- created 250 accounts in 16.434965 seconds
- created 250 accounts in 17.422578 seconds
- created 250 accounts in 16.413088 seconds
- created 250 accounts in 17.446779 seconds
- created 250 accounts in 16.428680 seconds
- created 250 accounts in 17.474911 seconds
- created 250 accounts in 17.498406 seconds
- created 250 accounts in 16.439295 seconds
- created 250 accounts in 17.503240 seconds
- created 250 accounts in 16.425590 seconds
- created 250 accounts in 17.493749 seconds
```

Is this the new loader code causing slower account creation, or a different reason?

simonhf commented:
Tried the loader account creation with the new follower using unsealed blocks, but it is still taking longer than the old loader per 250 accounts :-(

```
$ cat run-600-loader.sh.log | egrep "creating accounts" | perl -lane 'm~^([\d\.]+)~; ($ts)=($1); if(defined $ts_old){ printf qq[- created 250 accounts in %9.6f seconds\n], $ts - $ts_old; } $ts_old = $ts;'
- created 250 accounts in 17.479380 seconds
- created 250 accounts in 17.420236 seconds
- created 250 accounts in 17.455321 seconds
- created 250 accounts in 17.417838 seconds
- created 250 accounts in 17.493568 seconds
- created 250 accounts in 17.417041 seconds
- created 250 accounts in 16.458232 seconds
...
```

SaveTheRbtz commented:
Why is that a problem?  Previous loader was creating accounts w/ a speed of 5 accs/sec (50/10), the new one is at 14.7 accs/sec (250/17).  Even though there is a regression, the overall test runtime (for tests with >500 accounts) is reduced.

SaveTheRbtz commented:
I've changed the code to assume accounts being created on execution instead of on sealing.

SaveTheRbtz commented:
A side note:
Both logging and slow account creation are legit problems introduced in recent diffs but they are not a part of this diff.

simonhf commented:
> Why is that a problem? Previous loader was creating accounts w/ a speed of 5 accs/sec (50/10), the new one is at 14.7 accs/sec (250/17). Even though there is a regression, the overall test runtime (for tests with >500 accounts) is reduced.

The previous loader was creating 250 accounts in 10 seconds per me. And yes, the new one is 250 accounts in 17 seconds. This means at 220 TPS and the default multiplier of 50 then it'll create 11,000 accounts, and take an extra 11,000 / 250 * 7 = 308 seconds or 5.1 minutes extra to create those accounts. It's just boring delaying load start by an extra 5 minutes :-)

> I've changed the code to assume accounts being created on execution instead of on sealing.

Thanks! This seems to work. Here's a snippet of loader logs showing it's back to 10 seconds between creating batches of 250 accounts :-)

```
...
1655252035.065858 Wed Jun 15 00:13:55 2022 12:13AM INF creating accounts cumulative=6250 num=250 numberOfAccounts=6600
1655252035.341670 Wed Jun 15 00:13:55 2022 12:13AM DBG new block parsed blockHeight=260 blockID=22ab8fda0e45e80656e4dd631af961406b157a838223b073fd8126ea0716ac76 numCollections=0 numSeals=0 timeSinceLastBlock=900.588263 timeToGetBlockByHeight=0.67527 timeToParseBlock=0.692762 txsInBlock=0 txsInBlockUnknown=0 txsInProgress=0 txsTotal=25 txsTotalUnknown=25
1655252036.341069 Wed Jun 15 00:13:56 2022 12:13AM DBG new block parsed blockHeight=261 blockID=d066c738d838086d1aa73e45b9d8ed425e6c27a86c5234e17838b6ba86d8ebf3 numCollections=0 numSeals=0 timeSinceLastBlock=999.306964 timeToGetBlockByHeight=0.763592 timeToParseBlock=0.780462 txsInBlock=0 txsInBlockUnknown=0 txsInProgress=0 txsTotal=25 txsTotalUnknown=25
1655252037.241491 Wed Jun 15 00:13:57 2022 12:13AM DBG new block parsed blockHeight=262 blockID=7104494c4d209edc47fd3d26b26ca752a41bacfd9fedef4868d65fb821540e0d numCollections=0 numSeals=0 timeSinceLastBlock=900.320688 timeToGetBlockByHeight=0.692682 timeToParseBlock=0.711428 txsInBlock=0 txsInBlockUnknown=0 txsInProgress=0 txsTotal=25 txsTotalUnknown=25
1655252038.141346 Wed Jun 15 00:13:58 2022 12:13AM DBG new block parsed blockHeight=263 blockID=945d4d8a18744e1ce7f592293bbf058800a6ac175dc079420fa95b97f0bb1796 numCollections=0 numSeals=0 timeSinceLastBlock=899.718074 timeToGetBlockByHeight=0.953284 timeToParseBlock=0.966704 txsInBlock=0 txsInBlockUnknown=0 txsInProgress=0 txsTotal=25 txsTotalUnknown=25
1655252039.141632 Wed Jun 15 00:13:59 2022 12:13AM DBG new block parsed blockHeight=264 blockID=044c46f022e6b985dfb885c8341d1b9f62c652d05bd02d0a9a6e22ae89ca2bee numCollections=0 numSeals=0 timeSinceLastBlock=1000.190648 timeToGetBlockByHeight=0.609591 timeToParseBlock=0.626191 txsInBlock=0 txsInBlockUnknown=0 txsInProgress=0 txsTotal=25 txsTotalUnknown=25
1655252040.041726 Wed Jun 15 00:14:00 2022 12:14AM DBG new block parsed blockHeight=265 blockID=64312c96407e376c2cb4491e7b9a16ea6046d631bae3d735d20a4aba8a405b69 numCollections=0 numSeals=0 timeSinceLastBlock=900.02489 timeToGetBlockByHeight=0.732781 timeToParseBlock=0.746584 txsInBlock=0 txsInBlockUnknown=0 txsInProgress=0 txsTotal=25 txsTotalUnknown=25
1655252040.941442 Wed Jun 15 00:14:00 2022 12:14AM DBG new block parsed blockHeight=266 blockID=939936010f551c1e37e15ff6b0aa50619c27a325a5bfbe6a332496274e9c7f66 numCollections=0 numSeals=0 timeSinceLastBlock=899.599371 timeToGetBlockByHeight=0.665978 timeToParseBlock=0.681965 txsInBlock=0 txsInBlockUnknown=0 txsInProgress=0 txsTotal=25 txsTotalUnknown=25
1655252041.841138 Wed Jun 15 00:14:01 2022 12:14AM DBG new block parsed blockHeight=267 blockID=f47beef3c29bc494d79dd1b534feb9f0cd71540d45081ec3a31879422da69f63 numCollections=0 numSeals=0 timeSinceLastBlock=899.600821 timeToGetBlockByHeight=0.720797 timeToParseBlock=0.740765 txsInBlock=0 txsInBlockUnknown=0 txsInProgress=0 txsTotal=25 txsTotalUnknown=25
1655252043.342055 Wed Jun 15 00:14:03 2022 12:14AM DBG new block parsed blockHeight=268 blockID=0d1cb3e9b9f82ca7ae982d5febcde123d5388b15af1d45434312e9b1263f94a5 numCollections=1 numSeals=0 timeSinceLastBlock=1500.811127 timeToGetBlockByHeight=0.702482 timeToParseBlock=1.131622 txsInBlock=1 txsInBlockUnknown=1 txsInProgress=0 txsTotal=26 txsTotalUnknown=26
1655252043.741651 Wed Jun 15 00:14:03 2022 12:14AM DBG new block parsed blockHeight=269 blockID=40e4503f7dc2c4cc671cca5f21e246eb65083db8fa664723bd5e06fc319b4f42 numCollections=0 numSeals=0 timeSinceLastBlock=399.498838 timeToGetBlockByHeight=0.740371 timeToParseBlock=0.759386 txsInBlock=0 txsInBlockUnknown=0 txsInProgress=0 txsTotal=26 txsTotalUnknown=26
1655252044.142901 Wed Jun 15 00:14:04 2022 12:14AM INF creating accounts cumulative=6500 num=100 numberOfAccounts=6600
...
```

simonhf commented:
Suggestion: I'm assuming `timeSinceLastBlock=1002.67121` is the time in ms? In which case why not call it `msSinceLastBlock` or `timeMsSinceLastBlock` in order to make it more intention revealing? Same for other time metrics.

`$ cat run-600-loader.sh.log | egrep timeSinceLastBlock | tail -1`
`1655252668.445183 Wed Jun 15 00:24:28 2022 12:24AM DBG new block parsed blockHeight=896 blockID=c06339ebd8dc21343af4ee0f6612490ff645f1be997c6ba4bcc1e57d2a6c26d9 numCollections=3 numSeals=0 timeSinceLastBlock=1002.67121 timeToGetBlockByHeight=0.714712 timeToParseBlock=4.260045 txsInBlock=431 txsInBlockUnknown=0 txsInProgress=2868 txsTotal=128940 txsTotalUnknown=23942`

simonhf commented:
As a side note for @Kay-Zee, it looks like flow-api doesn't do any better job than flow-cli for grabbing the next block in a consistently predictable and shorter amount of time :-( Using this PR -- using flow-api instead of flow-cli -- to load test at 220 TPS on a 3 cluster localnet, then there are still large delays grabbing the next block, in this case up to 6+ seconds.

As a point of interest and note for @SaveTheRbtz, these types of delays can be much bigger on `mainnet` than the `localnet` delays shown here.

```
$ cat run-600-loader.sh.log | egrep timeSinceLastBlock | tail -1
1655252668.445183 Wed Jun 15 00:24:28 2022 12:24AM DBG new block parsed blockHeight=896 blockID=c06339ebd8dc21343af4ee0f6612490ff645f1be997c6ba4bcc1e57d2a6c26d9 numCollections=3 numSeals=0 timeSinceLastBlock=1002.67121 timeToGetBlockByHeight=0.714712 timeToParseBlock=4.260045 txsInBlock=431 txsInBlockUnknown=0 txsInProgress=2868 txsTotal=128940 txsTotalUnknown=23942

$ cat run-600-loader.sh.log | egrep timeSinceLastBlock | wc -l
896

$ cat run-600-loader.sh.log | egrep timeSinceLastBlock | perl -lane 'm~timeSinceLastBlock=(\d+)~; ($ms)=($1); while($ms =~ s~(\d)\d(x*)$~$1x$2~){} $h->{$ms}++; sub END{ foreach $ms(keys %{$h}){ printf qq[- %4s ms delay occurred %3d times\n], $ms, $h->{$ms}; } }' | sort
-    8 ms delay occurred   1 times
-   2x ms delay occurred   2 times
-   3x ms delay occurred   1 times
-   5x ms delay occurred   6 times
-   6x ms delay occurred   9 times
-   7x ms delay occurred  24 times
-   8x ms delay occurred  23 times
-   9x ms delay occurred 161 times
-  1xx ms delay occurred  80 times
-  2xx ms delay occurred   3 times
-  3xx ms delay occurred   3 times
-  4xx ms delay occurred   4 times
-  5xx ms delay occurred   2 times
-  6xx ms delay occurred   9 times
-  7xx ms delay occurred  24 times
-  8xx ms delay occurred 133 times
-  9xx ms delay occurred 159 times
- 1xxx ms delay occurred 153 times
- 2xxx ms delay occurred  51 times
- 3xxx ms delay occurred  31 times
- 4xxx ms delay occurred  12 times
- 5xxx ms delay occurred   4 times
- 6xxx ms delay occurred   1 times
```

SaveTheRbtz commented:
Used `InMS` suffix to match the rest of logging in the codebase, e.g. execution node's `timeSpentInMS`.

simonhf commented:
Validated that `-account-multiplier` is working and the `availableAccountsLo` log feature :-)

* Used the loader option `-account-multiplier 30` at 220 TPS to reduce the pre-created accounts to 6,600.
* After pre-creation the load ran for 600 seconds and the lowest available accounts was 2,609. Plenty of head room.

```
$ less run-600-loader.sh.log | egrep "availableAccountsLo"
1655252070.233269 Wed Jun 15 00:14:30 2022 12:14AM DBG discovered new account low availableAccountsLo=6580 numberOfAccounts=6600 workerID=0
1655252071.225118 Wed Jun 15 00:14:31 2022 12:14AM DBG discovered new account low availableAccountsLo=6289 numberOfAccounts=6600 workerID=0
1655252072.224839 Wed Jun 15 00:14:32 2022 12:14AM DBG discovered new account low availableAccountsLo=6032 numberOfAccounts=6600 workerID=0
1655252073.231090 Wed Jun 15 00:14:33 2022 12:14AM DBG discovered new account low availableAccountsLo=5752 numberOfAccounts=6600 workerID=0
1655252074.241186 Wed Jun 15 00:14:34 2022 12:14AM DBG discovered new account low availableAccountsLo=5634 numberOfAccounts=6600 workerID=0
1655252075.237110 Wed Jun 15 00:14:35 2022 12:14AM DBG discovered new account low availableAccountsLo=5320 numberOfAccounts=6600 workerID=0
1655252076.230017 Wed Jun 15 00:14:36 2022 12:14AM DBG discovered new account low availableAccountsLo=5117 numberOfAccounts=6600 workerID=0
1655252077.230181 Wed Jun 15 00:14:37 2022 12:14AM DBG discovered new account low availableAccountsLo=4880 numberOfAccounts=6600 workerID=0
1655252078.224413 Wed Jun 15 00:14:38 2022 12:14AM DBG discovered new account low availableAccountsLo=4798 numberOfAccounts=6600 workerID=0
1655252079.225929 Wed Jun 15 00:14:39 2022 12:14AM DBG discovered new account low availableAccountsLo=4413 numberOfAccounts=6600 workerID=0
1655252080.226505 Wed Jun 15 00:14:40 2022 12:14AM DBG discovered new account low availableAccountsLo=4217 numberOfAccounts=6600 workerID=0
1655252081.225369 Wed Jun 15 00:14:41 2022 12:14AM DBG discovered new account low availableAccountsLo=4159 numberOfAccounts=6600 workerID=0
1655252083.230202 Wed Jun 15 00:14:43 2022 12:14AM DBG discovered new account low availableAccountsLo=4050 numberOfAccounts=6600 workerID=0
1655252084.230650 Wed Jun 15 00:14:44 2022 12:14AM DBG discovered new account low availableAccountsLo=3728 numberOfAccounts=6600 workerID=0
1655252099.226284 Wed Jun 15 00:14:59 2022 12:14AM DBG discovered new account low availableAccountsLo=3569 numberOfAccounts=6600 workerID=0
1655252100.230247 Wed Jun 15 00:15:00 2022 12:15AM DBG discovered new account low availableAccountsLo=3357 numberOfAccounts=6600 workerID=0
1655252120.226676 Wed Jun 15 00:15:20 2022 12:15AM DBG discovered new account low availableAccountsLo=3312 numberOfAccounts=6600 workerID=0
1655252134.225289 Wed Jun 15 00:15:34 2022 12:15AM DBG discovered new account low availableAccountsLo=3261 numberOfAccounts=6600 workerID=0
1655252152.226881 Wed Jun 15 00:15:52 2022 12:15AM DBG discovered new account low availableAccountsLo=3085 numberOfAccounts=6600 workerID=0
1655252171.228749 Wed Jun 15 00:16:11 2022 12:16AM DBG discovered new account low availableAccountsLo=3057 numberOfAccounts=6600 workerID=0
1655252172.227575 Wed Jun 15 00:16:12 2022 12:16AM DBG discovered new account low availableAccountsLo=2766 numberOfAccounts=6600 workerID=0
1655252173.226975 Wed Jun 15 00:16:13 2022 12:16AM DBG discovered new account low availableAccountsLo=2609 numberOfAccounts=6600 workerID=0
```

simonhf commented:
Here we look at the debug output of the new loader in this PR, at the `txsInBlock` metric. By finding the first and last block during the load test, adding up all the transactions found in each block, and dividing by the timestamp delta between the first and last block, we can get the average output TPS for the entire load test, which is 219.6 for this load test with 220 input TPS... which looks right :-)

```
$ cat run-600-loader.sh.log | egrep txsInBlock | tail -1
1655252668.445183 Wed Jun 15 00:24:28 2022 12:24AM DBG new block parsed blockHeight=896 blockID=c06339ebd8dc21343af4ee0f6612490ff645f1be997c6ba4bcc1e57d2a6c26d9 numCollections=3 numSeals=0 timeSinceLastBlock=1002.67121 timeToGetBlockByHeight=0.714712 timeToParseBlock=4.260045 txsInBlock=431 txsInBlockUnknown=0 txsInProgress=2868 txsTotal=128940 txsTotalUnknown=23942

$ cat run-600-loader.sh.log | egrep txsInBlock | perl -lane 'if(m~^([\d\.]+) .* txsInBlock=(\d+)~){ ($ts,$tx)=($1,$2); if((not defined $load_start_ts) && ($tx > 1)){ $load_start_ts = $ts; } if(defined $load_start_ts){ $bx ++; $tx_total += $tx; $load_end_ts = $ts; printf qq[%d ], $tx; } } sub END{ $load_delta_ts = $load_end_ts - $load_start_ts; printf qq[<-- TX in each block\n- %d TX in %d blocks in %f seconds or %.1f average output TPS\n], $tx_total, $bx, $load_delta_ts, $tx_total / $load_delta_ts; }'
217 220 202 227 220 212 85 217 229 151 215 205 217 340 258 200 154 212 123 228 247 240 130 237 248 224 258 202 212 195 206 226 195 308 213 216 207 209 225 250 188 244 220 146 299 148 232 277 220 222 222 221 206 226 222 222 146 121 312 254 270 200 215 219 223 70 221 174 353 289 202 249 220 224 223 213 138 210 189 276 287 146 207 292 198 199 231 210 137 149 329 279 208 250 235 197 225 234 220 237 76 280 181 262 275 243 205 241 144 223 156 302 252 171 253 251 213 214 122 267 197 230 277 221 150 309 139 183 269 283 220 230 144 207 203 242 201 247 208 299 224 148 148 280 184 263 291 223 133 168 236 294 181 217 289 145 241 133 315 269 240 130 232 299 74 288 217 296 220 137 157 167 250 378 220 226 152 138 219 241 361 222 212 154 59 295 319 234 234 172 137 171 337 308 207 158 159 205 250 257 301 234 139 224 200 292 167 300 135 201 216 177 287 217 230 161 290 141 280 150 293 141 311 199 222 310 147 144 295 165 268 225 198 209 167 300 290 154 143 303 136 277 231 242 195 132 324 225 283 190 226 161 285 221 58 227 272 233 265 77 248 342 165 83 273 185 386 151 144 202 282 371 150 218 162 224 314 165 175 96 320 303 202 242 142 225 357 152 79 268 221 276 262 67 306 221 281 210 83 292 203 268 248 142 237 220 296 145 195 212 312 192 194 216 328 197 203 129 237 208 324 139 205 236 201 309 213 161 274 177 208 345 77 231 282 87 286 301 228 129 241 234 279 146 214 210 263 87 285 153 409 138 226 218 238 174 236 226 216 194 240 233 414 0 283 173 126 0 299 137 581 319 241 141 137 286 220 225 141 298 283 144 171 203 258 177 142 334 261 131 146 386 202 173 302 191 246 137 0 144 286 72 513 454 158 78 293 347 77 164 356 222 140 165 316 201 59 332 196 233 147 319 221 210 206 162 389 142 65 370 206 104 370 211 291 78 289 217 228 144 340 232 77 306 172 342 67 299 146 210 229 290 303 77 260 242 197 310 122 212 295 134 206 401 93 153 384 198 147 198 295 143 172 359 136 245 207 231 203 150 324 75 349 209 224 193 141 417 232 126 265 220 142 250 262 84 385 226 60 316 273 139 232 250 251 106 168 356 76 139 439 86 320 233 76 348 177 212 194 125 487 78 120 270 0 165 333 0 653 149 200 331 153 242 300 268 83 228 169 292 208 265 150 305 205 79 314 148 297 230 199 382 138 198 304 150 238 293 144 356 150 155 298 206 81 374 141 192 289 176 197 250 207 190 214 264 219 204 207 148 206 297 201 358 0 147 0 76 78 431 <-- TX in each block
- 128912 TX in 590 blocks in 587.001623 seconds or 219.6 average output TPS
```

simonhf reviewed (approved):
LGTM :-)

Tomorrow I'll continue investigating a workaround for download all traces from Tempo... :-)

SaveTheRbtz commented:
bors merge
